### PR TITLE
Move `Write.Tx` modules to `cardano-balance-tx`

### PR DESCRIPTION
**Pull request by Anviking**
- [x] Factor out `std-gen-seed` library
- [x] Move `Write.Tx` modules to `cardano-balance-tx` 

### Comments

- Unit tests to be moved in separate PR and/or task.

### Issue Number

ADP-3170

### CODE REVIEW

File: lib/balance-tx/cardano-balance-tx.cabal [hunk 1]
```diff
@@ -40,7 +40,33 @@ library
...
+    , std-gen-seed
+    , cardano-ledger-mary
+    , cardano-ledger-babbage
+    , cardano-ledger-alonzo
+    , cardano-ledger-allegra
+    , cardano-crypto-class
+    , ouroboros-consensus-cardano
+    , int-cast
+    , cardano-addresses
+    , bytestring
```
Comment (jonathanknowles) [inline lib/balance-tx/cardano-balance-tx.cabal]: Is it possible to sort these dependencies into alphabetical order? 🙏🏻 
(It's easier to spot duplicates that way)

File: lib/wallet/src/Cardano/Wallet/Transaction.hs [hunk 2]
```diff
@@ -122,6 +120,8 @@ import qualified Cardano.Api as Cardano
 import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
 import qualified Cardano.Wallet.Primitive.Types.Tx.TxOut as TxOut
 import qualified Cardano.Wallet.Write.Tx as Write
+import Cardano.Wallet.Write.Tx.SizeEstimation
+    ( TxWitnessTag )
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Transaction.hs]: Wrong place ^^^

File: lib/std-gen-seed/std-gen-seed.cabal [hunk 3]
```diff
@@ -0,0 +1,64 @@
+cabal-version:   3.0
+name:            std-gen-seed
+version:         0.1.0.0
+synopsis:        modules to support applications for the cardano wallet
```
Comment (jonathanknowles) [inline lib/std-gen-seed/std-gen-seed.cabal]:
```suggestion
synopsis:        Support for standard random number generator seeds
```

File: cabal.project [hunk 4]
```diff
@@ -76,6 +76,7 @@ packages:
   , lib/wai-middleware-logging/
   , lib/wallet-benchmarks/
   , lib/wallet/
+  , lib/std-gen-seed
```
Comment (jonathanknowles) [inline cabal.project]: Should this have a trailing slash:

```suggestion
  , lib/std-gen-seed/
```

File: lib/balance-tx/lib/Cardano/Wallet/Write/Tx/SizeEstimation.hs [hunk 5]
```diff
@@ -697,3 +701,14 @@ mkLedgerTxOut txOutEra address bundle =
...
       where
         txOut = W.TxOut address bundle
+
+data TxWitnessTag
+    = TxWitnessByronUTxO
+    | TxWitnessShelleyUTxO
+    deriving (Show, Eq)
+
+assumedTxWitnessTag :: UTxOAssumptions -> TxWitnessTag
+assumedTxWitnessTag = \case
```
Comment (Anviking) [inline lib/balance-tx/lib/Cardano/Wallet/Write/Tx/SizeEstimation.hs:711]: Why not leave it in the UtxoAssumptions module?

File: lib/std-gen-seed/std-gen-seed.cabal [hunk 6]
```diff
@@ -0,0 +1,64 @@
+cabal-version:   3.0
+name:            std-gen-seed
+version:         0.1.0.0
```
Comment (jonathanknowles) [inline lib/std-gen-seed/std-gen-seed.cabal]: If we're going to start at the beginning, then we can start at `0.0.0.0`:

```suggestion
version:         0.0.0.0
```

File: lib/std-gen-seed/std-gen-seed.cabal [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+library
+  default-language:   Haskell2010
+  default-extensions:
+    NoImplicitPrelude
+    DerivingStrategies
+    OverloadedStrings
+
+  ghc-options:
+    -O2 -Wall -Wincomplete-record-updates -Wincomplete-uni-patterns
```
Comment (jonathanknowles) [inline lib/std-gen-seed/std-gen-seed.cabal]: We should probably not specify `-O2` unless it's a release build.

File: lib/std-gen-seed/std-gen-seed.cabal [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+test-suite unit
+  default-language:
+      Haskell2010
+  default-extensions:
+      NoImplicitPrelude
+      OverloadedStrings
+  ghc-options:
+      -threaded -rtsopts
+      -Wall -O2 -Werror
```
Comment (jonathanknowles) [inline lib/std-gen-seed/std-gen-seed.cabal]: Similar to earlier comment, we should probably not specify `-O2` here unless it's for a release build.

File: lib/wallet/cardano-wallet.cabal [hunk 9]
```diff
@@ -794,6 +781,7 @@ test-suite unit
     , mtl
     , network
     , network-uri
+    , std-gen-seed
```
Comment (jonathanknowles) [inline lib/wallet/cardano-wallet.cabal]: We should put this in the alphabetical order... :D

File: lib/wallet/cardano-wallet.cabal [hunk 10]
```diff
@@ -1069,6 +1056,7 @@ benchmark db
     , cardano-wallet-primitive
     , cardano-wallet-read
     , cardano-wallet-test-utils
+    , cardano-balance-tx
```
Comment (jonathanknowles) [inline lib/wallet/cardano-wallet.cabal]: Likewise (alphabetical order)

**Review by jonathanknowles — COMMENTED** [review]
Hi @Anviking 

Just some preliminary comments!

File: lib/balance-tx/lib/Cardano/Wallet/Write/Tx/SizeEstimation.hs (hunk 5 above)
Comment (Anviking) [reply lib/balance-tx/lib/Cardano/Wallet/Write/Tx/SizeEstimation.hs:711]: 🤔 `TxWitnessTag` is ultimately only needed for `SizeEstimation.hs`-functions, and seems more internal to me. I'd favor not putting it in `Write.UTxOAssumption` now.

**Review by jonathanknowles — APPROVED** [review]

**Pull request merged**